src/iceberg/type.h

```diff
@@ -0,0 +1,149 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+/// \file iceberg/type.h
+/// Data types for Iceberg.
+
+#include <memory>
+#include <span>
+#include <string>
+#include <vector>
+
+#include "iceberg/iceberg_export.h"
+#include "iceberg/type_fwd.h"
+
+namespace iceberg {
+
+/// \brief A type combined with a name.
+class ICEBERG_EXPORT Field {
+ public:
+  /// \brief Construct a field.
+  /// \param[in] field_id The field ID.
+  /// \param[in] name The field name.
+  /// \param[in] type The field type.
+  /// \param[in] optional Whether values of this field are required or nullable.
+  Field(int32_t field_id, std::string name, std::shared_ptr<Type> type, bool optional);
+
+  /// \brief Construct an optional (nullable) field.
+  static Field MakeOptional(int32_t field_id, std::string name,
+                            std::shared_ptr<Type> type);
+  /// \brief Construct a required (non-null) field.
+  static Field MakeRequired(int32_t field_id, std::string name,
+                            std::shared_ptr<Type> type);
+
+  /// \brief Get the field ID.
+  [[nodiscard]] int32_t field_id() const;
+
+  /// \brief Get the field name.
+  [[nodiscard]] std::string_view name() const;
+
+  /// \brief Get the field type.
+  [[nodiscard]] const std::shared_ptr<Type>& type() const;
+
+  /// \brief Get whether the field is optional.
+  [[nodiscard]] bool optional() const;
+
+  /// \brief Get a user-readable string representation of the field.
+  [[nodiscard]] std::string ToString() const;
+
+  /// \brief Compare two fields for equality.
+  [[nodiscard]] bool Equals(const Field& other) const;
+
+  friend bool operator==(const Field& lhs, const Field& rhs) { return lhs.Equals(rhs); }
+
+  friend bool operator!=(const Field& lhs, const Field& rhs) { return !(lhs == rhs); }
+};
+
+/// \brief Interface for a data type for a field.
+class ICEBERG_EXPORT Type {
+ public:
+  virtual ~Type() = default;
+
+  /// \brief Get the type ID.
+  [[nodiscard]] virtual TypeId type_id() const = 0;
+
+  /// \brief Get a user-readable string representation of the type.
+  [[nodiscard]] virtual std::string ToString() const = 0;
+
+  /// \brief Compare two types for equality.
+  [[nodiscard]] virtual bool Equals(const Type& other) const = 0;
+
+  /// \brief Is this a primitive type (may not have child fields)?
+  [[nodiscard]] virtual bool is_primitive() const = 0;
+
+  /// \brief Is this a nested type (may have child fields)?
```

Comment:
What raise my interests is that:
1. Does iceberg has a "void" ( or null ) bottom type?
2. Does iceberg allowed to define an empty structure?

Reply:
There is `unknown`. 

I don't think the spec says anything either way about an empty struct. I didn't check any implementations. If allowed, you could think of it as a `void` type.

Reply:
The `UnknownType` will be part of the V3 specification.

An empty struct is pretty common (for example, an unpartitioned table has an empty struct as a partition).